[lexical-code] Bug Fix: 2 Tabs on single line selection should indent

jeromew opened the pull request:
## Description

There is a bug in the Tab mechanism in lexical-code.
When you select a single line (full selection), a Tab indents and tests are ok.
The problem is that if you Tab a second time, then the selection is removed instead of indenting a second time.

This PR fixes the bug by making sure that after a Tab in such a scenario, the newly inserted TabNode is added to the selection.
The PR also modifies the [lexical-code] unit tests by making sure to test both forward selection & backward selection of the line

Closes #7541

jeromew commented on packages/lexical-code/src/CodeHighlighter.ts:
```diff
@@ -610,11 +610,32 @@ function $handleMultilineIndent(type: LexicalCommand<void>): boolean {
+    const tabNode = $createTabNode();
     if ($isLineBreakNode(firstOfLine)) {
-      firstOfLine.insertAfter($createTabNode());
+      firstOfLine.insertAfter(tabNode);
     } else {
-      firstOfLine.insertBefore($createTabNode());
+      firstOfLine.insertBefore(tabNode);
     }
+    // The added tabNode should be included in the current selection.
+    // Otherwise, the selection will not cover the full line and a new
+    // Tab Keypress would remove the selection via INSERT_TAB_COMMAND
+    const backward = selection.isBackward();
+    const focusKey = backward ? 'anchor' : 'focus';
+    let [anchorNode, anchorOffset]: [TextNode, number] = [tabNode, 0];
+    let [focusNode, focusOffset]: [TextNode, number] = [
+      selection[focusKey].getNode() as TextNode,
+      selection[focusKey].offset,
+    ];
+    if (backward) {
+      [anchorNode, focusNode] = [focusNode, anchorNode];
+      [anchorOffset, focusOffset] = [focusOffset, anchorOffset];
+    }
+    selection.setTextNodeRange(
+      anchorNode,
+      anchorOffset,
+      focusNode,
+      focusOffset,
+    );
```
Thanks your strategy indeed seems better. I have tried to follow it but am blocked by an error in the type system that I don't understand how to fix :

```js
error TS2339: Property 'clone' does not exist on type 'CodeHighlightNode | TabNode | LineBreakNode'.
      Property 'clone' does not exist on type 'CodeHighlightNode'.
    614    const firstOfLineClone = firstOfLine.clone(firstOfLine);
```
I checked everywhere I could, and clone seems defined in flow scripts on all three classes. 

Here it complains about 'CodeHighlightNode' if I run `npm run tsc` I get

```js
error TS2339: Property 'clone' does not exist on type 'LineBreakNode | TabNode | CodeHighlightNode'.
  Property 'clone' does not exist on type 'LineBreakNode'.
614    const firstOfLineClone = firstOfLine.clone(firstOfLine);
```
which complains about 'LineBreakNode'

does it ring a bell for you ?

jeromew replied:
(another practical question) do you prefer PR squashed or with additional commits ?

jeromew replied:
maybe my problem comes from the fact that `clone` is a static method (?)

etrepum replied:
You should never call clone directly. There’s no need to do any sort of cloning or copying for this operation. https://lexical.dev/docs/concepts/node-cloning

etrepum replied:
No need to squash or rebase. Squash merge is used so PRs end up as a single commit on main.

jeromew replied:
ok sorry I had misunderstood the `replace` api.
now I think I understand the basics of your proposed strategy : try to let the core handle the manipulation of the selection with an algo like

```js
if ($isLineBreakNode(firstOfLine)) {
      firstOfLine.insertAfter($createTabNode());
      } else {
      const tabNode = $createTabNode();
      firstOfLine.replace(tabNode);
      tabNode.insertAfter(firstOfLine);
    }
```

I don't know exactly if this was the idea, but my problem is that this works when the line has > 1 nodes.
But it does not work when the line has exactly 1 node because from what I see insertAfter never extends the selection in this case.
the tests for example, have only 1 node : "function", and they break because of this problem.

can you tell me if I misunderstood your proposed strategy or if you were thinking of something else / another API that I could use to keep the algo simple ? thanks

etrepum replied:
Right, it looks like you will have to explicitly set either the anchor or the focus of the selection to also handle the single node case. That was the alternate strategy I suggested

etrepum reviewed (commented):
It looks like some of the CodeBlock tests are failing https://github.com/facebook/lexical/actions/runs/15107985447/job/42504150622?pr=7544

```
  3 failed
    [chromium] › packages/lexical-playground/__tests__/e2e/CodeBlock.spec.mjs:399:3 › CodeBlock › Can maintain indent when creating new lines 
    [chromium] › packages/lexical-playground/__tests__/e2e/CodeBlock.spec.mjs:491:3 › CodeBlock › Can (un)indent multiple lines at once 
    [chromium] › packages/lexical-playground/__tests__/e2e/CodeBlock.spec.mjs:1155:3 › CodeBlock › When pressing CMD/Ctrl + Left, CMD/Ctrl + Right, the cursor should go to the start of the code 
```

jeromew commented:
@etrepum I started looking at the cross effects between tabulation & Indentation via the playground menu.

it will need a bit more work to find a good strategy.

For now, can we agree that there is a bug in `$isSelectionInCode` (https://github.com/facebook/lexical/blob/main/packages/lexical-code/src/CodeHighlighter.ts#L471)

the code currently checks if anchor&focus are the same CodeNode or share the same CodeNode as their respective parent, but it does not check cases that exist in the wild where for example anchor is a CodeNode and anchor.is(focus.getParent()).

From the name of the function, I expected it to return true whenever the selection is located under the same CodeNode so I don't really know if the current implementation should be considered as a feature or as a bug.

I could maybe replace
```js
function $isSelectionInCode(selection: null | BaseSelection): boolean {
  if (!$isRangeSelection(selection)) {
    return false;
  }
  const anchorNode = selection.anchor.getNode();
  const focusNode = selection.focus.getNode();
  if (anchorNode.is(focusNode) && $isCodeNode(anchorNode)) {
    return true;
  }
  const anchorParent = anchorNode.getParent();
  return $isCodeNode(anchorParent) && anchorParent.is(focusNode.getParent());
}
```

by

```js
function $isSelectionInCode(selection: null | BaseSelection): boolean {
  if (!$isRangeSelection(selection)) {
    return false;
  }
  const anchorNode = selection.anchor.getNode();
  const anchorCodeNode = $isCodeNode(anchorNode) ? anchorNode : anchorNode.getParent();
  const focusNode = selection.focus.getNode();
  const focusCodeNode = $isCodeNode(focusNode) ? focusNode : focusNode.getParent();
  return anchorCodeNode.is(focusCodeNode);
}
```

I also have a question regarding `$getCodeLines` which currently extracts **non-empty** line sections within a selection.

The problem I have with it is that we lose the information about empty lines, and this is why we end up in a "// Just one line" section in cases where there can be multiple lines like "\n\n\nHELLO\n\n" and no easy way to discriminate between the cases.

I would like to return empty lines as empty arrays is `$getCodeLines` to keep information about the structure of the code block.

I think it could be possible because both `$getCodeLines` and `$isSelectionInCode` seem to be only used inside `CodeHighlighter.ts`

I am going to try these modifications on my local machine and see where it takes me but I would be interested in your feedback regarding these 2 functions. Thanks !

jeromew commented:
ok I did only check e2e CodeBlocks locally. Let me check.

jeromew commented:
@etrepum is there something I should to to be able to see the Vercel build logs on github ? I created an account, but I get a 404 when I click on the failed deployment. thank you

etrepum commented:
the vercel build that failed can mostly be reproduced with `npm run -w packages/lexical-website build`

```
../lexical-code/src/CodeHighlighter.ts:492:5 - error TS18047: 'maybeAnchorCodeNode' is possibly 'null'.

492     maybeAnchorCodeNode.is(maybeFocusCodeNode) &&
```

A faster fail for this one is the integrity check (which doesn't run before admin approval since you are a first time contributor)

```
npm run ci-check
```

etrepum commented on packages/lexical-code/src/CodeHighlighter.ts:
```diff
@@ -468,43 +469,65 @@ function isEqual(nodeA: LexicalNode, nodeB: LexicalNode): boolean {
+    : focusNode.getParent();
+
+  return (
+    maybeAnchorCodeNode.is(maybeFocusCodeNode) &&
+    $isCodeNode(maybeAnchorCodeNode)
```
```suggestion
    $isCodeNode(maybeAnchorCodeNode) &&
    maybeAnchorCodeNode.is(maybeFocusCodeNode)
```

The type guard should go before calling methods that require it to be true

jeromew commented:
ok i see i have a flow/tsc issue in the "Lexical Tests" ; I am going to fix it after the e2e-tests finish

etrepum commented:
Some of the collab tests are flaky, I restarted that test and suspect it will pass with no changes.

I don't know which code editor was used as inspiration for the behavior of this particular node's behavior, but I suspect that modeling VSCode is probably what most people would expect at this point from a web based code editor.

jeromew commented:
The current behavior of the PR simply happened to pass all tests with a simplification of the code and was not blocking for me. but I preferred to ask you as it could have been felt as a regression. I just observed that different editors have different corner case behaviors & Tab for "delete intra single-line selection & replace with Tab" is a feature I have never used (I use DEL to delete a selection) so I imagined lowering its usage would not be a problem.

I agree the default VSCode behavior could be taken as an example. I don't know if VSCode has 

I'll try first to document the features of Indent/Outdent/Tab regarding effect on the content of the code editor & the effect on the selection.
Then I'll manually test the status of lexical on these features.
then I'll try to add tests for these and make the necessary modifications to lexical so it passes them.

etrepum reviewed (commented):
I think the primary difference between this behavior and VSCode behavior is that VSCode will not indent a line if no characters are selected in that line (e.g. in a forward selection the focus is at offset 0 of the first text node in a line)

for example (`|` used to denote selection start/end):

Before tab
```
|line
|another line
```

After tab (vscode)
```
|  line
|another line
```

After tab (this PR)
```
|  line
  |another line
```

The VSCode behavior is preferable I think

etrepum reviewed (approved):
I think what we can do is merge this as-is since it seems to be a net improvement in the ways that matter, and matches the existing behavior. The behavior change to match VSCode indent can happen in a separate isolated PR

jeromew commented:
ok. thanks @etrepum.
I'll open another issue to document and track the status of the whole Indent/Outdent/Tab feature in CodeNode

jeromew commented:
> The VSCode behavior is preferable I think

yes I agree. To be honest I thought I had fixed that.
the whole thing needs more unit tests

I added a test for the
```
|line1
|

line2
```
case which does

```
    |line1
|

line2
```

but did not think about adding your case !
